functions/webhooks/twitter/FlexToTwitter.protected.ts

```diff
@@ -0,0 +1,127 @@
+import Twit from 'twit';
+import '@twilio-labs/serverless-runtime-types';
+import { Context, ServerlessCallback } from '@twilio-labs/serverless-runtime-types/types';
+import {
+  responseWithCors,
+  bindResolve,
+  error400,
+  error500,
+  success,
+  send,
+} from '@tech-matters/serverless-helpers';
+
+type EnvVars = {
+  TWITTER_CONSUMER_KEY: string;
+  TWITTER_CONSUMER_SECRET: string;
+  TWITTER_ACCESS_TOKEN: string;
+  TWITTER_ACCESS_TOKEN_SECRET: string;
+  CHAT_SERVICE_SID: string;
+};
+
+export type Body = {
+  recipientId?: string;
+  Source?: string;
+  Body?: string;
+  From?: string;
+  ChannelSid?: string;
+  EventType?: string;
+};
+
+const sendTwitterMessage = async (
+  context: Context<EnvVars>,
+  recipientId: string,
+  messageText: string,
+) => {
+  const T = new Twit({
+    consumer_key: context.TWITTER_CONSUMER_KEY,
+    consumer_secret: context.TWITTER_CONSUMER_SECRET,
+    access_token: context.TWITTER_ACCESS_TOKEN,
+    access_token_secret: context.TWITTER_ACCESS_TOKEN_SECRET,
+    timeout_ms: 60 * 1000, // optional HTTP request timeout to apply to all requests.
+    strictSSL: true, // optional - requires SSL certificates to be valid.
+  });
+
+  const sendTo = {
+    event: {
+      type: 'message_create',
+      message_create: {
+        target: {
+          recipient_id: recipientId,
+        },
+        message_data: {
+          text: messageText,
+        },
+      },
+    },
+  };
+
+  // @ts-ignore
+  return T.post('direct_messages/events/new', sendTo);
+};
+
+export const handler = async (
+  context: Context<EnvVars>,
+  event: Body,
+  callback: ServerlessCallback,
+) => {
+  const response = responseWithCors();
+  const resolve = bindResolve(callback)(response);
+
+  try {
+    const { recipientId, Body } = event;
+
+    if (recipientId === undefined) {
+      resolve(error400('recipientId'));
+      return;
+    }
+
+    if (Body === undefined) {
+      resolve(error400('Body'));
+      return;
+    }
+
+    console.log('------ FlexToTwitter excecution ------');
+
+    if (event.Source === 'SDK') {
+      const TwitResponse = await sendTwitterMessage(context, recipientId, Body);
+      console.log('Message sent from SDK call: ', Body);
+      resolve(success(TwitResponse));
+      return;
+    }
+
+    if (event.Source === 'API' && event.EventType === 'onMessageSent') {
```

Comment:
I'm a little confused with `event.Source.` When it's value is `SDK` does it means it's a message that a counselor typed in herself, but when it's `API` it could mean it came from the autopilot? It's not clear to me which scenarios it's `SKD` or `API`.

Reply:
Yes, you get it right, counselor messages sent from Flex have SDK source, and every other message I've seen so far is API, including messages that the DM create in behalf of the client getting in contact (the child), so here we let them all go thru except the ones sent by the client, to avoid "redirecting self messages".